### PR TITLE
InMemoryLazySparseMerkleTree

### PR DESCRIPTION
**Pull request by DanieleDiBenedetto**
Java classes and JNI functions Rust implementation

### CODE REVIEW

File: jni/src/main/java/com/horizen/merkletreenative/InMemoryLazySparseMerkleTree.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    public boolean isPositionEmpty(long position){
+        if (merkleTreePointer == 0)
+            throw new IllegalStateException("InMemoryLazySparseMerkleTree instance was freed.");
+        return nativeIsPositionEmpty(position);
+    }
+
+    private native FieldElement nativeAddLeaves(PositionLeaf[] leaves);
+
+    // Add leaves to tree, compute and return the root
+    public FieldElement addLeaves(List<PositionLeaf> leaves){
```
Comment (i-Alex) [inline jni/src/main/java/com/horizen/merkletreenative/InMemoryLazySparseMerkleTree.java:39]: Why we should return root?
I would expect to add leaves and return boolean if it was successful

File: api/src/lib.rs [hunk 2]
```diff
@@ -3318,4 +3349,218 @@ ffi_export!(
...
+        Err(e) => throw!(&_env, "java/lang/Exception", format!("Cannot insert leaves: {}", e.to_string()).as_str(), JObject::null().into_inner())
+    }    
+});
+
+ffi_export!(
+    fn Java_com_horizen_merkletreenative_InMemoryLazySparseMerkleTree_nativeRemoveLeaves(
+    _env: JNIEnv,
+    _tree: JObject,
+    _positions: jlongArray,
+) -> jobject
```
Comment (i-Alex) [inline api/src/lib.rs:3476]: Java part expect `void` as a result.
I think, we don't need to calculate the root as a last step. Only remove leafs if possible.

**Review by i-Alex — CHANGES_REQUESTED** [review]

**Review by i-Alex — APPROVED** [review]
Perfect!

**Comment by Yousefatm999** [comment]
Good

**Review by Yousefatm999 — APPROVED** [review]

**Review by albertog78 — APPROVED** [review]

**Review by i-Alex — APPROVED** [review]

**Review by albertog78 — APPROVED** [review]

**Pull request merged**